src/h3lib/lib/h3Index.c

```diff
@@ -779,6 +779,86 @@ void H3_EXPORT(h3ToGeoBoundary)(H3Index h3, GeoBoundary* gb) {
                           H3_EXPORT(h3IsPentagon)(h3), gb);
 }
 
+/**
+ * Returns the max number of possible icosahedron faces an H3 index
+ * may intersect.
+ *
+ * @return int count of faces
+ */
+int H3_EXPORT(maxFaceCount)(H3Index h3) {
+    // a pentagon always intersects 5 faces, a hexagon never intersects more
+    // than 2 (but may only intersect 1)
+    return H3_EXPORT(h3IsPentagon)(h3) ? 5 : 2;
+}
+
+/**
+ * Find all icosahedron faces intersected by a given  H3 index, represented
+ * as integers from 0-19. The array is sparse; since 0 is a valid value,
+ * invalid array values are represented as -1. It is the responsibility of
+ * the caller to filter out invalid values.
+ *
+ * @param h3 The H3 index
+ * @param out Output array. Must be of size maxFaceCount(h3).
+ */
+void H3_EXPORT(h3GetFaces)(H3Index h3, int* out) {
+    int res = H3_GET_RESOLUTION(h3);
+    int isPentagon = H3_EXPORT(h3IsPentagon)(h3);
+
+    // We can't use the vertex-based approach here for class II pentagons,
+    // because all their vertices are on the icosahedron edges. Their
+    // direct child pentagons cross the same faces, so use those instead.
+    if (isPentagon && !isResClassIII(res)) {
+        H3Index childPentagon = makeDirectChild(h3, 0);
+        H3_EXPORT(h3GetFaces)(childPentagon, out);
+        return;
+    }
+
+    // convert to FaceIJK
+    FaceIJK fijk;
+    _h3ToFaceIjk(h3, &fijk);
+
+    // Get all vertices as FaceIJK addresses. For simplicity, always
+    // initialize the array with 6 verts, ignoring the last one for pentagons
+    FaceIJK fijkVerts[NUM_HEX_VERTS];
+    int vertexCount;
+
+    if (isPentagon) {
+        vertexCount = NUM_PENT_VERTS;
+        _faceIjkPentToVerts(&fijk, &res, fijkVerts);
+    } else {
+        vertexCount = NUM_HEX_VERTS;
+        _faceIjkToVerts(&fijk, &res, fijkVerts);
+    }
+
+    // We may not use all of the slots in the output array,
+    // so fill with invalid values to indicate unused slots
+    int faceCount = H3_EXPORT(maxFaceCount)(h3);
+    for (int i = 0; i < faceCount; i++) {
+        out[i] = INVALID_FACE;
+    }
+
+    // add each vertex face, using the output array as a hash set
+    for (int i = 0; i < vertexCount; i++) {
+        FaceIJK vert = fijkVerts[i];
```

Comment:
Nit: This will copy the vertex definition and then perform the operations below, which probably isn't what you wanted. This could be `FaceIJK* vert = &fijkVerts[i];` then `_adjustPentVertOverage(vert, res);`, `_adjustOverageClassII(vert, res, 0, 1);`, and `int face = vert->face;`

Reply:
Makes sense, will update.